File: 438_FindAllAnagramsInAString.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(p) > len(s): return []
        countP = [0] * 26
        countS = [0] * 26
        output = []
        n = len(p)
        for i in p:
            countP[ord(i)-97] += 1
        for i in range(len(p)):
            countS[ord(s[i])-97] += 1
        if countP == countS: output.append(0)
        for i in range(1,len(s)-n+1):
            countS[ord(s[i-1])-97] -= 1
            countS[ord(s[i+n-1])-97] += 1
            if countS == countP:
                output.append(i)
        return output

File: C++/438_FindAllAnagramsInAString.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> countS(26,0), countP(26,0), output;
        if (p.size() > s.size()) return output;
        for(int i = 0; i < p.size(); ++i)
        {
            ++countP[p[i]-'a'];
            ++countS[s[i]-'a'];
        }
        if (countP == countS) output.push_back(0);
        for (int x = p.size(); x < s.size(); ++x){
            ++countS[s[x]-'a'];
            --countS[s[x-p.size()]-'a'];
            if (countS == countP){
                output.push_back(x-p.size()+1);
            }
        }
        return output;
    }
};